jinahub/indexers/storage/RedisStorage/redis_storage.py

```diff
@@ -0,0 +1,125 @@
+    :param default_batch_size: default batch size
+    :param args: other arguments
+    :param kwargs: other keyword arguments
+    """
+    def __init__(self,
+                 hostname: str = '0.0.0.0',
+                 # default port on linux
+                 port: int = 6379,
+                 db: int = 0,
+                 default_traversal_paths: Tuple = ('r',),
+                 default_batch_size: int = 32,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.default_batch_size = default_batch_size
+        self.default_traversal_paths = default_traversal_paths
+        self.hostname = hostname
+        self.port = port
+        self.db = db
+        self.connection_pool = redis.ConnectionPool(host=self.hostname, port=self.port, db=self.db)
+        self.logger = JinaLogger(self.__class__.__name__)
+
+    def get_query_handler(self) -> 'Redis':
+        """Get the redis client handler.
+        """
+        import redis
+        try:
+            r = redis.Redis(connection_pool=self.connection_pool)
+            r.ping()
+            return r
+        except redis.exceptions.ConnectionError as r_con_error:
+            self.logger.error('Redis connection error: ', r_con_error)
+            raise
+
+    def _query_batch(self, docs: Iterable[Document]):
+        with self.get_query_handler() as redis_handler:
+            results = redis_handler.mget([doc.id for doc in docs])
+            for doc, result in zip(docs, results):
+                if not result:
+                    continue
+                data = bytes(result)
+                retrieved_doc = Document(data)
+                # TODO: check why is this performed for postgres storage
+                # retrieved_doc.pop('embedding')
+                doc.MergeFrom(retrieved_doc)
+
+    @requests(on='/search')
+    def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        """Searches documents in the redis server by document ID
+
+        :param docs: document array
+        :param parameters: parameters to the request
+        """
+        if docs:
+            document_batches_generator = get_docs_batch_generator(
+                docs,
+                traversal_path=parameters.get('traversal_paths', self.default_traversal_paths),
+                batch_size=parameters.get('batch_size', self.default_batch_size)
+            )
+            for document_batch in document_batches_generator:
+                self._query_batch(document_batch)
+
+    def _upsert_batch(self, docs: Iterable[Document]):
+        with self.get_query_handler().pipeline() as redis_handler:
+            redis_handler.mset({doc.id: doc.SerializeToString() for doc in docs})
+            redis_handler.execute()
+
+    @requests(on=['/index', '/update'])
+    def upsert(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        """Inserts documents in the redis server where the key is the document ID. If a document with the same ID
```

Comment:
Not sure it should behave in an upsert way. The others don't. I'd rather keep it predictable across the executors. 

Can u make index index, and update only update if there are docs with those ids? If there aren't you don't update them

Reply:
done